### ISSUE TITLE
Performance of ModernMT v4

### ISSUE THREAD

**Issue by nicolabertoldi** [modernmt/modernmt#485]
We have compared performance of ModernMT v4 against modernMT v3.3, exploiting the corresponding default configuration.
The comparison has been performed on a small/medium-sized public benchmark, namely the [Multilingual Task of the IWSLT-2017 Evaluation](https://sites.google.com/site/iwsltevaluation2017/TED-tasks) for the **Italian-English** direction.

On the official test set, the two systems achieve the following BLEU scores and decoding times (seconds/sentence):

system | BLEU | decoding time
------------ | ------------- | -------------
ModernMT v3.3 | 38.2 | 0.22
ModernMT v4    | 38.9 | 0.14

**Note**:
- training and evaluation were run on an Ubuntu on a commodity machine having one GPU (GeForce GTX 1080 Ti with 11Gb) and a 8-core CPU (i7-7700 CPU with 32 Gb RAM) on board.

If you intend to replicate these results or compare with them, here are a few useful info.

The benchmark data are available in this archive [benchmark_iwslt2017.tar.gz](https://github.com/modernmt/modernmt/files/3226538/benchmark_iwslt2017.tar.gz).

v3.3 system was trained with the default configuration and stop after 30K steps as follows:
```
mmt create it en ./train --train-steps 30000
```

v4 system was trained with the default configuration as follows:
```
mmt create it en ./train
```

Both v3.3 and v4 system have been evaluated as follows:
```
mmt evaluate it en --path ./test
```

#### Benchmark preparation
For the sake of completeness, here is the description how the archive [benchmark_iwslt2017.tar.gz](https://github.com/modernmt/modernmt/files/3226538/benchmark_iwslt2017.tar.gz) has been prepared.

First, download the training data from https://wit3.fbk.eu/mt.php?release=2017-01-trnmted (click the link in the table), and the test data from https://wit3.fbk.eu/mt.php?release=2017-01-mted-test (click on both en-it and it-en links in the table).

Then,  extract raw data as follows:
```
mkdir train test
tar xzf DeEnItNlRo-DeEnItNlRo.tgz DeEnItNlRo-DeEnItNlRo/train.tags.it-en.it
tar xzf DeEnItNlRo-DeEnItNlRo.tgz DeEnItNlRo-DeEnItNlRo/train.tags.it-en.en
tar zxf it-en.tgz
tar zxf en-it.tgz

cat DeEnItNlRo-DeEnItNlRo/train.tags.it-en.it | grep -v "^ *<" | sed 's#^ *##g'  > train/train.it
cat DeEnItNlRo-DeEnItNlRo/train.tags.it-en.en | grep -v "^ *<" | sed 's#^ *##g'  > train/train.en
cat it-en/IWSLT17.TED.tst2017.mltlng.it-en.it.xml | grep '<seg' | sed 's# *<\/*seg[^>]*> *##g' > test/tst2017.it
cat en-it/IWSLT17.TED.tst2017.mltlng.en-it.en.xml | grep '<seg' | sed 's# *<\/*seg[^>]*> *##g' > test/tst2017.en
```
**Note:**
- for training, all lines starting with a xml tag are removed.
- for test, only lines starting with the xml tag `seg` are taken into account 
- the training archive also contains two development sets (dev2010 and test2010), but they were not used in our system and not included in the archive.

**Comment by Hasan0ff** [comment]
Hey! Great work. I've been training my model for 9days and 20hours+ (and it's still on-going)

`epoch 028:  49300 / 210564 loss=4.523, nll_loss=3.014, ppl=8.08, wps=9502, ups=10, wpb=953.969, bsz=42.035, num_updates=9.3493e+06, lr=2.06843e-06, gnorm=8.645, clip=0.000, oom=0.000, , wall=4946, train_wall=793566`

I had used --debug so I stopped the model to check for the quality and there isn't much difference in quality with 2-3 day training and the current 9+ days training. The dataset is about 3GB per language. Moreover, it's been over 3 days that the loss has gone from 4.7 to 4.5. Is there any reason for such slow training?

**Comment by EtienneAb3d** [comment]
@Hasan0ff, possibly you started your training before this: https://github.com/modernmt/modernmt/issues/478#issuecomment-496492404

**Comment by Hasan0ff** [comment]
> @Hasan0ff, possibly you started your training before this: [#478 (comment)](https://github.com/modernmt/modernmt/issues/478#issuecomment-496492404)

does it mean that i have to reinstall and re-train? (there is no problem as i'm ready to do it if thats what it takes)

**Comment by nicolabertoldi** [comment]
@Hasan0ff,


ModernMT v3.X did not include any termination policy, but the maximum number of train-steps, which was by default 1000000. Hence, it is formally "correct" that your training run for so long time.

Instead, new release v4 implements by default a smart termination policy, which stops automatically the training process in the estimated best moment. It's also possible to impose a fixed number of training steps; in this case the automatic termination policy is disabled.


Regarding the performance, it is difficult to say which is the best point to stop the training, because it strongly depends from the size of the training data, the size of the model, the complexity of the domain, number and power of the GPUs employed, and so on. In our experiments on a 8-GPU machine, we observed that 3/4 hours is fine to reach the best performance for a `transformer_mmt_base` architecture with about 4M training words, while 4/5 days is fine  for a `transformer_mmt_big` architecture with about 2G training words.
Extrapolating from these figures, I can say that in your case 2/3 days can be enough on a similar machine.

We also observed that after the potentially optimal stop point, the performance tends to be stable or slightly decrease. This behaviour depends again on the training and data characteristics.


Finally, I would kindly ask you for the future to open a new issue more focused on your actual problem; comments to existing issues should be posted only if related to them.

**Comment by nicolabertoldi** [comment]
@Hasan0ff 
models trained with v3.X are not compatible with v4.

Models trained with v4 (beta version) before the introduction of the smart termination policy are compatible with the current release (v4).

**Comment by Hasan0ff** [comment]
@nicolabertoldi alright. so i will stop the training and re-install with the current release instead of using beta. nevertheless, great job on the beta version as it was very good too and sorry for mixing my issue with the topic.
great job and please keep improving modernmt!

**Comment by mzeidhassan** [comment]
Hi @nicolabertoldi 
Thanks for this useful comparison. I have one question.  Did you use transformer_big or transformer_base in MMT v3.3? 

If you used both variations, did you notice any difference in terms of BLEU score?

Thanks!

**Comment by nicolabertoldi** [comment]
@mzeidhassan 
For both v3.3 and v4 I used the corresponding default configurations. i.e. "transformer_base"  architecture.

I dd not run the comparison using a transformer_big architecture.

**Comment by mzeidhassan** [comment]
Thanks @nicolabertoldi! I think it's worth-trying transformer_big because it usually has a better BLEU score. 
I haven't tested MMT v4 yet, but I hope other users will share their experiences in this issue. Thanks again for making MMT v4 available.
